mono/utils/atomic.h

```diff
@@ -24,158 +24,162 @@ F/MonoDroid( 1568): shared runtime initialization error: Cannot load library: re
-/* MSVC before 2013 only defines InterlockedAdd* for the Itanium architecture */
-static inline gint32 InterlockedAdd(volatile gint32 *dest, gint32 add)
+static inline gint64
+mono_atomic_inc_i64 (volatile gint64 *dest)
 {
-	return InterlockedExchangeAdd (dest, add) + add;
+	return InterlockedIncrement64 ((LONG64 volatile *)dest);
 }
-#endif
 
-#if defined(_MSC_VER) && !defined(InterlockedAdd64)
-#if defined(InterlockedExchangeAdd64)
-/* This may be defined only on amd64 */
-static inline gint64 InterlockedAdd64(volatile gint64 *dest, gint64 add)
+static inline gint32
+mono_atomic_dec_i32 (volatile gint32 *dest)
 {
-	return InterlockedExchangeAdd64 (dest, add) + add;
+	return InterlockedDecrement ((LONG volatile *)dest);
 }
-#else
-static inline gint64 InterlockedAdd64(volatile gint64 *dest, gint64 add)
-{
-	gint64 prev_value;
-
-	do {
-		prev_value = *dest;
-	} while (prev_value != InterlockedCompareExchange64(dest, prev_value + add, prev_value));
 
-	return prev_value + add;
+static inline gint64
+mono_atomic_dec_i64 (volatile gint64 *dest)
+{
+	return InterlockedDecrement64 ((LONG64 volatile *)dest);
 }
-#endif
-#endif
 
-#ifdef HOST_WIN32
-#define TO_INTERLOCKED_ARGP(ptr) ((volatile LONG*)(ptr))
-#else
-#define TO_INTERLOCKED_ARGP(ptr) (ptr)
-#endif
-
-/* And now for some dirty hacks... The Windows API doesn't
- * provide any useful primitives for this (other than getting
- * into architecture-specific madness), so use CAS. */
+static inline gint32
+mono_atomic_xchg_i32 (volatile gint32 *dest, gint32 exch)
+{
+	return InterlockedExchange ((LONG volatile *)dest, (LONG)exch);
+}
 
-static inline gint32 InterlockedRead(volatile gint32 *src)
+static inline gint64
+mono_atomic_xchg_i64 (volatile gint64 *dest, gint64 exch)
 {
-	return InterlockedCompareExchange (TO_INTERLOCKED_ARGP (src), 0, 0);
+	return InterlockedExchange64 ((LONG64 volatile *)dest, (LONG64)exch);
 }
 
-static inline gint64 InterlockedRead64(volatile gint64 *src)
+static inline gpointer
+mono_atomic_xchg_ptr (volatile gpointer *dest, gpointer exch)
 {
-	return InterlockedCompareExchange64 (src, 0, 0);
+	return InterlockedExchangePointer ((PVOID volatile *)dest, (PVOID)exch);
 }
 
-static inline gpointer InterlockedReadPointer(volatile gpointer *src)
+
+static inline gint32
+mono_atomic_fetch_add_i32 (volatile gint32 *dest, gint32 add)
 {
-	return InterlockedCompareExchangePointer (src, NULL, NULL);
+	return InterlockedExchangeAdd ((LONG volatile *)dest, (LONG)add);
 }
 
-static inline void InterlockedWrite(volatile gint32 *dst, gint32 val)
+
+static inline gint64
+mono_atomic_fetch_add_i64 (volatile gint64 *dest, gint64 add)
 {
-	InterlockedExchange (TO_INTERLOCKED_ARGP (dst), val);
+	return InterlockedExchangeAdd64 ((LONG64 volatile *)dest, (LONG)add);
 }
 
-static inline void InterlockedWrite64(volatile gint64 *dst, gint64 val)
+static inline gint8
+mono_atomic_load_i8 (volatile gint8 *src)
 {
-	InterlockedExchange64 (dst, val);
+	return *src;
```

Comment:
Does something else need done here?

Reply:
I had the same thoughts, but since this was the current implementation since 4 years back, I kept it as is. On x86 and x64 Windows there should not be any risk of reading partial memory if properly aligned and using the volatile keyword should prevent the compiler to rearrange the instructions (/volatile:ms extension). Having that said, I took a look at C++11 atomics implementation on Windows for char and short types (even for 32-bit int's and on x64 64-bit ints), and by default it is using sequential consistent memory access and that actually inserts a explicit compiler barrier, _ReadWriteBarrier (no CPU MemoryBarrier though). If you use the lazy memory model it is implemented as above, just as a regular mov. Since the _ReadWriteBarrier only prevent the compiler to reorder load/stores, it should do what /volatile:ms does, but since the compiler also supports /volatile:iso (used as default on ARM), I believe the library explicitly added the compiler barrier to just have an implementation independent of /volatile compiler switch. We could add the same semantics as the C++11 atomics implementation have, just to be compliant, but the above should still be OK on x86, x64 Windows architectures. I update the PR with the change.